master/pyext/src/validation/__init__.py

```diff
@@ -12,527 +12,623 @@
+        atoms = {}
+        before_atom_site = []
+        after_atom = []
+        for i, j in enumerate(all_lines):
+            if len(j) > 0 and '_atom_site.' in j[0]:
+                if len(before_atom_site) == 0:
+                    before_atom_site = all_lines[:i+1]
+                atom_site[i] = j[0]
+            elif ('_atom_site.B_iso_or_equiv' not in list(atom_site.values())) and len(list(atom_site.values())) > 0:
+                atom_site[i] = '_atom_site.B_iso_or_equiv'
+            elif ('_atom_site.occupancy' not in list(atom_site.values())) and len(list(atom_site.values())) > 0:
+                atom_site[i] = '_atom_site.occupancy'
+
+            elif ('_atom_site.label_seq_id' not in list(atom_site.values())) and len(list(atom_site.values())) > 0:
+                atom_site[i] = '_atom_site.label_seq_id'
+
+        total_list = list(atom_site.values())
+        index_biso = total_list.index('_atom_site.B_iso_or_equiv')
+        index_occu = total_list.index('_atom_site.occupancy')
+        index_label_seq = total_list.index('_atom_site.label_seq_id')
+        for i, j in enumerate(all_lines):
+            if len(j) > 0 and ('ATOM' in j[0] or 'HETATM' in j[0]) and (i > list(atom_site.keys())[-1]):
+                if len(j) <= index_occu:
                     j.extend(['1'])
-                elif j[index_occu]=='.':
-                    j[index_occu]='0.67'
-                if len(j)<=index_biso:
+                elif j[index_occu] == '.':
+                    j[index_occu] = '0.67'
+                if len(j) <= index_biso:
                     j.extend(['1'])
-                elif j[index_biso]=='.':
-                    j[index_biso]='0.00'
-                atoms[i]=j
-            elif len(j)> 0 and  (i > list(atom_site.keys())[-1]):
-                if len(after_atom)==0:
-                    after_atom=all_lines[i:]
-        return before_atom_site,atom_site,atoms,after_atom
+                elif j[index_biso] == '.':
+                    j[index_biso] = '0.00'
+                if len(j) <= index_label_seq:
+                    j.extend(['1'])
+                elif j[index_label_seq] == '.':
+                    j[index_label_seq] = str(i)
+                atoms[i] = j
+            elif len(j) > 0 and (i > list(atom_site.keys())[-1]):
+                if len(after_atom) == 0:
+                    after_atom = all_lines[i:]
+        return before_atom_site, atom_site, atoms, after_atom
 
     def rewrite_mmcif(self):
         """ This function writes a temporary mmcif file that can be parsed by molprobity
         after checking occupancy and b-iso parameters """
-        before_atom_site,atom_site,atoms,after_atom=self.mmcif_get_lists()
+        before_atom_site, atom_site, atoms, after_atom = self.mmcif_get_lists()
         if os.path.isfile('test.cif'):
             os.remove('test.cif')
-        file_re=open('test.cif','w')
+        file_re = open('test.cif', 'w')
+        forbidden = ['°', 'µ', 'Å', '"']
```

Comment:
I think you're trying to coerce your text here to ASCII. Python strings can do this for you already. I would probably just use something like

    with open('dodgy-non-ascii-file', 'r', encoding='ascii', errors='replace') as fh:
        ascii_contents = fh.read()

If you already have the non-ascii data in a Python string, just use `ascii_string = non_ascii_string.encode('ascii', errors='replace')`

`errors='replace'` will map all of the forbidden characters (anything where `ord(chr) > 127` in this case) to `?`.